core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/commit/CommitInfoSearchRequest.java

```diff
@@ -88,11 +100,100 @@ protected CommitInfos createEmptyResult(int limit) {
 		return new CommitInfos(limit, 0);
 	}
 	
-	private void addBranchClause(final ExpressionBuilder builder) {
+	private void addSecurityFilter(final ExpressionBuilder builder, RepositoryContext context) {
+		
+		final User user = context.service(User.class);
+		if (user.isAdministrator() || user.hasPermission(Permission.requireAll(Permission.OPERATION_BROWSE, Permission.ALL))) {
+			if (containsKey(OptionKey.BRANCH)) {
+				addBranchClause(builder, context, null, null, true);		
+			}
+			return;
+		}
+		
+		final List<Permission> readPermissions = user.getPermissions().stream()
+				.filter(p -> Permission.ALL.equals(p.getOperation()) || Permission.OPERATION_BROWSE.equals(p.getOperation()))
+				.collect(Collectors.toList());
+		
+		final Set<String> exactResourceIds = readPermissions.stream()
+				.flatMap(p -> p.getResources().stream())
+				.filter(resource -> !resource.endsWith("*"))
+				.collect(Collectors.toSet());
+		
+		final Set<String> resourceIdPrefixes = readPermissions.stream()
+				.flatMap(p -> p.getResources().stream())
+				.filter(resource -> resource.endsWith("*"))
+				.map(resource -> resource.substring(0, resource.length() - 1))
+				.collect(Collectors.toSet());
+		
 		if (containsKey(OptionKey.BRANCH)) {
-			final Collection<String> branchPaths = getCollection(OptionKey.BRANCH, String.class);
-			builder.filter(branches(branchPaths));
+			addBranchClause(builder, context, exactResourceIds, resourceIdPrefixes, false);
+			return;
 		}
+				
+		SetView<String> resourceIds = Sets.union(exactResourceIds, resourceIdPrefixes);
+		ResourceRequests.prepareSearch()
+			.filterByIds(resourceIds)
+			.setLimit(resourceIds.size())
+			.setFields(ResourceDocument.Fields.ID, 
+					ResourceDocument.Fields.BRANCH_PATH,
+					ResourceDocument.Fields.RESOURCE_TYPE)
+			.buildAsync()
+			.getRequest()
+			.execute(context)
+			.stream()
+			.filter(TerminologyResource.class::isInstance)
+			.map(TerminologyResource.class::cast)
+			.forEach(r -> {
+				if (resourceIdPrefixes.contains(r.getId())) {
+					final String branchPattern = String.format("%s(/[a-zA-Z0-9.~_\\-]{3,100})?", r.getBranchPath());
+					builder.filter(regexp(BRANCH, branchPattern));
+				}
+			});		
+	}
+	
+	private void addBranchClause(final ExpressionBuilder builder, RepositoryContext context, 
+			Set<String> exactResourceIds, Set<String> resourceIdPrefixes, boolean isAdmin) {		
+		final Collection<String> branchPaths = getCollection(OptionKey.BRANCH, String.class);
+				
+		ExpressionBuilder branchFilter = Expressions.builder();
+		ResourceSearchRequestBuilder resourceRequest = ResourceRequests.prepareSearch();
+		
+		if (isAdmin) {
+			resourceRequest.setLimit(Integer.MAX_VALUE);			
```

Comment:
This is just going to be too much. Why do you want to return all resources in case the user is an admin?
In that case there is no need to filter anything just apply the incoming filter value and skip the security filter.